src/executor/read_only_runtime_host.rs

```diff
@@ -0,0 +1,389 @@
+                    // relatively expensive operation (in the order of milliseconds).
+                    // While it could be tempting to use a system cache, this function is expected
+                    // to be called only right before runtime upgrades. Considering that runtime
+                    // upgrades are quite uncommon and that a caching system is rather non-trivial
+                    // to set up, the approach of recompiling every single time is preferred here.
+                    // TODO: number of heap pages?! we use the default here, but not sure whether that's correct or if we have to take the current heap pages
+                    let vm_prototype = match host::HostVmPrototype::new(
+                        req.wasm_code(),
+                        executor::DEFAULT_HEAP_PAGES,
+                        vm::ExecHint::Oneshot,
+                    ) {
+                        Ok(w) => w,
+                        Err(_) => {
+                            self.vm = req.resume(Err(()));
+                            continue;
+                        }
+                    };
+
+                    match super::core_version(vm_prototype) {
+                        Ok((version, _)) => {
+                            // TODO: optimize
+                            self.vm = req.resume(Ok(&parity_scale_codec::Encode::encode(&version)));
+                        }
+                        Err(_) => {
+                            self.vm = req.resume(Err(()));
+                        }
+                    }
+                }
+
+                host::HostVm::StartStorageTransaction(tx) => {
+                    self.vm = tx.resume();
+                }
+
+                host::HostVm::EndStorageTransaction { resume, rollback } => {
+                    if rollback {
+                        todo!() // TODO:
+                    }
+
+                    self.vm = resume.resume();
+                }
+
+                host::HostVm::LogEmit(req) => {
+                    // We add a hardcoded limit to the logs generated by the runtime in order to
+                    // make sure that there is no memory leak. In practice, the runtime should
+                    // rarely log more than a few hundred bytes. This limit is hardcoded rather
+                    // than configurable because it is not expected to be reachable unless
+                    // something is very wrong.
+                    // TODO: optimize somehow? don't create an intermediary String?
+                    let message = req.to_string();
+                    if self.logs.len().saturating_add(message.len()) >= 1024 * 1024 {
+                        return RuntimeHostVm::Finished(Err(Error::LogsTooLong));
+                    }
+
+                    self.logs.push_str(&message);
+                    self.vm = req.resume();
+                }
+
+                _ => {
+                    panic!("Forbidden");
```

Comment:
```suggestion
                    RuntimeHostVm::Finished(Err(Error::ForbiddenHostCall))
```

(add the `ForbiddenHostCall` variant to `Error`)